provisioning_client/adapters/hsm_client_riot.c

```diff
@@ -155,33 +136,49 @@ static const HSM_CLIENT_X509_INTERFACE x509_interface =
     hsm_client_riot_get_common_name
 };
 
-static int generate_root_ca_info(HSM_CLIENT_X509_INFO* riot_info, RIOT_ECC_SIGNATURE* tbs_sig)
+// Free the mbedtls_mpi members of the signature
+static void ecc_signature_destroy(RIOT_ECC_SIGNATURE* tbs_sig)
+{
+    mbedtls_mpi_free(&tbs_sig->r);
+    mbedtls_mpi_free(&tbs_sig->s);
+}
+
+static void x509_cert_free(RIOT_ECC_PUBLIC* pub, RIOT_ECC_PRIVATE* priv)
+{
+    mbedtls_ecp_point_free(pub);
+    mbedtls_mpi_free(priv);
+}
+
+static int generate_root_ca_info(HSM_CLIENT_X509_INFO* riot_info)
 {
     int result;
     uint8_t der_buffer[DER_MAX_TBS] = { 0 };
     DERBuilderContext der_ctx = { 0 };
     DERBuilderContext der_pri_ctx = { 0 };
     RIOT_STATUS status;
-
-    memcpy(&riot_info->ca_root_pub, eccRootPubBytes, sizeof(ecc_publickey));
-    memcpy(&riot_info->ca_root_priv, eccRootPrivBytes, sizeof(ecc_privatekey));
+    RIOT_ECC_SIGNATURE tbs_sig = { 0 };
 
     // Generating "root"-signed DeviceID certificate
     DERInitContext(&der_ctx, der_buffer, DER_MAX_TBS);
     DERInitContext(&der_pri_ctx, der_buffer, DER_MAX_TBS);
-
-    if (X509GetDeviceCertTBS(&der_ctx, &X509_ROOT_TBS_DATA, &riot_info->ca_root_pub) != 0)
+    if ((status = RiotCrypt_DeriveEccKey(&riot_info->ca_root_pub, &riot_info->ca_root_priv,
```

Comment:
What is g_digest?  There are no special requirements or expectations on the root CA key.  So it isn't **wrong** to create a derived key here based on g_digest, whatever it is.  But it does suggest there is a prescribed method or requirements for a root CA key.

Even adding a comment here is fine; something to note that for dev/test we don't care about the root key bytes and the root CA key in production should follow all the usual best practices.

Reply:
Also the label is incorrect.  There isn't an existing label defined for this key type because it is only needed on a call to derive the key (and the root CA key isn't normally derived in this way).  If you elect to keep the DeriveEccKey call for the root key, then you should add something like #define RIOT_LABEL_ROOTCA "ROOTCAKEY" somewhere for the label.

Reply:
```C
static int g_digest_initialized = 0;
static uint8_t g_digest[DICE_DIGEST_LENGTH] = { 0 };
static unsigned char g_uds_seed[DICE_UDS_LENGTH] = {
    0x54, 0x10, 0x5D, 0x2E, 0xCD, 0x07, 0xF9, 0x01,
    0x99, 0xB3, 0x95, 0xC7, 0x42, 0x61, 0xA0, 0x8C,
    0xFF, 0x27, 0x1A, 0x0D, 0xF6, 0x6F, 0x1F, 0xE0,
    0x00, 0x34, 0xBB, 0x11, 0xF7, 0x98, 0x9A, 0x12 };
static uint8_t g_CDI[DICE_DIGEST_LENGTH] = {
    0x91, 0x75, 0xDB, 0xEE, 0x90, 0xC4, 0xE1, 0xE3,
    0x74, 0x47, 0x2C, 0x8A, 0x55, 0x3F, 0xD2, 0xB8,
    0xE9, 0x79, 0xEE, 0xF1, 0x62, 0xF8, 0x64, 0xDA,
    0x50, 0x69, 0x4B, 0x3E, 0x5A, 0x1E, 0x3A, 0x6E };
```
the g_digest in question is hashed here ln 406:
```C
    else if ((status = RiotCrypt_Hash(g_digest, RIOT_DIGEST_LENGTH, g_CDI, DICE_DIGEST_LENGTH)) != RIOT_SUCCESS)
```
it is then used to derive the device_id_pub/priv keys
then is hashed again here ln 418:
```C
    // Combine CDI and FWID, result in digest
    else if ((status = RiotCrypt_Hash2(g_digest, DICE_DIGEST_LENGTH, g_digest, DICE_DIGEST_LENGTH, firmware_id, RIOT_DIGEST_LENGTH)) != RIOT_SUCCESS)
```
it is then used to derive the alias_key_pub/priv keys
However I'm not sure what the digest is actually suppose to emulate @jebrando might be able to explain more in depth

Reply:
A comment would probably be helpful, will add, I don't think the define label is needed since it is stored in the x509 info tag struct ln 96 - 128:
```C
typedef struct HSM_CLIENT_X509_INFO_TAG
{
    // In Riot these are call the device Id pub and pri
    RIOT_ECC_PUBLIC     device_id_pub;
    RIOT_ECC_PRIVATE    device_id_priv;

    RIOT_ECC_PUBLIC     alias_key_pub;
    RIOT_ECC_PRIVATE    alias_key_priv;

    RIOT_ECC_PUBLIC     ca_root_pub;
    RIOT_ECC_PRIVATE    ca_root_priv;
```
However if that is the route that is required then I can memset the define to the struct member

Reply:
Baked-in Dev CA Root, no longer deriving - Fixed